lib/spack/spack/package.py

```diff
@@ -1552,104 +1562,154 @@ def _sanity_check_extension(self):
+        Make sure that extension is usable by extendee in given fileystem view
+        without modifying the install directory of either package. This method
+        should both link all necessary files and ensure the extension is
+        usable. For example: All python extensions need to be added to the
+        corresponding `easy_install.pth` file.
 
         Package authors can override this method to support other
         extension mechanisms.  Spack internals (commands, hooks, etc.)
         should call do_activate() method so that proper checks are
         always executed.
 
         """
+        if not path_view:
+            prefix = self.prefix
+        else:
+            prefix = path_view
 
         def ignore(filename):
             return (filename in spack.store.layout.hidden_file_paths or
                     kwargs.get('ignore', lambda f: False)(filename))
 
         tree = LinkTree(extension.prefix)
-        conflict = tree.find_conflict(self.prefix, ignore=ignore)
+        conflict = tree.find_conflict(prefix, ignore=ignore)
         if conflict:
             raise ExtensionConflictError(conflict)
 
-        tree.merge(self.prefix, ignore=ignore)
+        tree.merge(prefix, ignore=ignore)
 
     def do_deactivate(self, **kwargs):
-        """Called on the extension to invoke extendee's deactivate() method."""
+        """Called on the extension to invoke extendee's deactivate() method.
+
+        `path_view` can be specified to deactivate package in given fileystem
+        view.
+
+        `remove_dependets=True` deactivates extensions depending on this
+        package instead of raising an error.
+        """
         self._sanity_check_extension()
         force = kwargs.get('force', False)
+        path_view = kwargs.get("path_view", "")
+        verbose = kwargs.get("verbose", True)
+        remove_dependents = kwargs.get("remove_dependents", False)
 
         # Allow a force deactivate to happen.  This can unlink
         # spurious files if something was corrupted.
         if not force:
             spack.store.layout.check_activated(
-                self.extendee_spec, self.spec)
+                self.extendee_spec, self.spec, path_view=path_view)
 
             activated = spack.store.layout.extension_map(
-                self.extendee_spec)
+                self.extendee_spec, path_view=path_view)
             for name, aspec in activated.items():
                 if aspec == self.spec:
                     continue
                 for dep in aspec.traverse(deptype='run'):
                     if self.spec == dep:
-                        msg = ("Cannot deactivate %s because %s is activated "
-                               "and depends on it.")
-                        raise ActivationError(
-                            msg % (self.spec.short_spec, aspec.short_spec))
-
-        self.extendee_spec.package.deactivate(self, **self.extendee_args)
+                        if remove_dependents:
```

Comment:
Got it - I overlooked the fact that it was actually deactivating the dependents. I'm assuming this falls into the category of changes that you made that weren't essential to the core goal of this PR but annoyed you while you were working on it. The recursive deactivations initially made me worried but this looks OK. I'd like to see a test for it (although in general there's some discussion to be had about testing).